### PR TITLE
feat(auto-rules): JavaScript Rule Match Expressions

### LINKED ISSUE #413: Document automatic rules API
Automatic rules are being implemented piece-by-piece across multiple PRs, and the API will be in flux for some time. Once it is settled, or close to it, API docs must be updated before the new API can be included in a release.

See #355

### LINKED ISSUE #540: Auto rules should offer matching on more than mainClass/targetAlias
Rules should be able to match targets by more criteria than simply "same alias".

Examples of rule matches that should be possible:

- [ ] "all targets"
- [ ] "target has connectUrl 'service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi'"
- [ ] "target has alias Foo"
- [ ] "target has main class io.cryostat.Cryostat"
- [ ] "target is listening on port 9999"
- [ ] "target has label app=myproject"
- [ ] "target has annotation cryostat.io/target=true"
- [ ] "target has main class io.cryostat.Cryostat AND target has annotation cryostat.io/profiling=true"
- [ ] "target has main class io.cryostat.Cryostat OR target has main class com.example.App"

Related: #280 #495 #413 #523

### PR DESCRIPTION
**Pull request by andrewazores**
Depends on #523

Fixes #540
Closes #413

### CODE REVIEW

File: src/main/java/io/cryostat/rules/RuleMatcher.java
Comment (ebaron) [inline src/main/java/io/cryostat/rules/RuleMatcher.java:138]: Are these events meant to be expanded in the future? Right now it seems like the BindingsCreationEvent always occurs at the same time as the RuleAppliesEvent, but doesn't provide any additional info.

**Review by ebaron — COMMENTED** [review]
Looking good! Just a couple of comments.

File: src/main/java/io/cryostat/rules/MatchExpressionValidator.java
Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:81]: I'm anonymously extending `SimpleTreeVisitorES5_1`, which is a concrete class implementing the `TreeVisitor` interface, and only overriding the visitor methods that we want to disallow. All other visitor methods are implemented by the `SimpleTreeVisitorES5_1` base class and inherited by this anonymous extension class, which implicitly pass those AST nodes by default. So, any visitor methods for node types not overridden by this class are implicitly passed, and all the ones I have explicitly named by overriding the relevant visitors are failed. Do you mean to replace this with a class where all of the visitor methods are directly implemented by the class and there is a composition delegate `SimpleTreeVisitor` field for passing, so that all the AST node types/visitor methods are enumerated in the implementation, rather than having the implicit passes that I have here?

File: src/main/java/io/cryostat/rules/RuleMatcher.java
Comment (andrewazores) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:138]: The `RuleAppliesEvent` starts strictly before and ends strictly after the `BindingsCreationEvent`, with a 1:1 relationship. I meant for these events to be used for profiling how long Nashorn evaluation of match expressions takes, and also to see how much of that time is actually evaluation by the rule engine vs. time spent in Cryostat preparing the bindings to pass to the engine. The time taken by Nashorn evaluation is roughly the duration of the `RuleAppliesEvent` minus the duration of the `BindingsCreationEvent`.

File: src/main/java/io/cryostat/rules/RuleMatcher.java
Comment (jan-law) [inline src/main/java/io/cryostat/rules/RuleMatcher.java:65]: Why does `createBindings()` need to be called each time the script engine wants to compare a `matchExpression` to the `serviceRef`'s annotations?

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:65]: That's a good question/observation. The bindings need to be created to take the full `ServiceRef` structure and slightly simplify it before passing it into the script engine (ie turn `AnnotationKeys` just into Strings), and both the `matchExpression` and the `serviceRef` have constant state that doesn't change over time, so the `Bindings` object would be a good candidate for caching/memoization. I'm not sure how expensive it actually is to create the bindings however, so it may not be worth the overhead of adding a cache and dealing with cache invalidation when `ServiceRefs` disappear. The `BindingsCreationEvent` could help to profile this and figure out if it's worthwhile.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:65]: We could also simply cache the result of whether a given rule's `matchExpression` applies to a given `ServiceRef`, instead of caching the `Binding`, and make it even faster by bypassing the script engine entirely, actually. If it is worth doing any caching then this would be the smarter way to go about it. This still incurs extra overhead in needing to do cache invalidation when `Rule` definitions are deleted or `ServiceRef`s disappear, though.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:65]: ... but then again, how often do we actually re-check a Rule against a ServiceRef? When a new Rule is added that rule alone is checked against all known ServiceRefs at the time, and vice-versa when a new ServiceRef appears it is checked against all of the Rules. So in either scenario, these are all new/unique combinations of Rule/ServiceRef, so the cache would be empty for all of these. Would we end up needing to re-evaluate it?

So maybe caching the ServiceRef -> Binding does make the most sense, since at least that would help in the case of a new Rule appearing. After the first one, we already have the Binding for each ServiceRef (assuming the set of ServiceRefs hasn't changed), so we can pass the same binding to the script engine to evaluate against the new rule matchExpression.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:65]: Anyway, tl;dr, my hunch is that the majority of the time is spent actually setting up the script engine, passing data to it, executing it, and marshalling the results back into Java, and comparatively little time is spent creating the Binding object. If that's true then it probably isn't worth caching the Bindings.

File: src/main/java/io/cryostat/rules/MatchExpressionValidator.java
Comment (ebaron) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:81]: Right, so that nothing is implicitly passed. I'm thinking it might be slightly less error prone to maintain, but I'll leave it up to you.

File: src/main/java/io/cryostat/rules/RuleMatcher.java
Comment (ebaron) [reply src/main/java/io/cryostat/rules/RuleMatcher.java:138]: Ah, I understand now. I hadn't considered it being a duration thing.

File: src/main/java/io/cryostat/rules/MatchExpressionValidator.java
Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:81]: Okay, that's fair enough. I'll refactor it to do as you suggest. It's also probably worth pulling it into its own class, too.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:81]: Hmm - this doesn't seem to work as expected, actually. The `TreeVisitor` only seems to have one of its methods called, and then execution resumes at the line after `cut.accept(treeVisitor, rule.getMatchExpression())`. I expected its methods to be called repeatedly as the parser proceeded, but that doesn't seem to be how it works. The first node type passed is the CompilationUnit, so it will pass everything after that point. I also tried iterating over `cut.getSourceElements()`, but this just gives me a list of AST children of the CompilationUnit, which can also be other generally acceptable abstract node types. So to get the "explicit passing to a delegate" piece working, it seems like I would need to recursively descend the whole AST myself - doesn't seem worth it.

I'll extract the TreeVisitor to a top-level class of its own at least, but I'll leave it with the original implementation of only observing fail-worthy AST nodes.

File: src/main/java/io/cryostat/rules/MatchExpressionValidator.java
Comment (hareetd) [inline src/main/java/io/cryostat/rules/MatchExpressionValidator.java:55]: If the match expression were to be improperly formatted, would `cut.accept(...)` throw an `IllegalMatchExpressionException` due to one of the overridden visit methods in the `MatchExpressionTreeVisitor` being called?

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:55]: Probably - it would depend on what "improperly formatted" looks like. `MatchExpressionValidatorTest` exercises various failure modes.

File: src/main/java/io/cryostat/rules/RuleMatcher.java
Comment (hareetd) [inline src/main/java/io/cryostat/rules/RuleMatcher.java:65]: Does `scripEngine.eval(...)` return an `Object` because we don't know whether the script we passed in actually evaluates to a `boolean` (i.e. it could be executing other types of code)?

File: src/test/java/io/cryostat/rules/RuleProcessorTest.java
Comment (hareetd) [inline src/test/java/io/cryostat/rules/RuleProcessorTest.java:241]: Is it optional for the target alias match expression to be of the form `target.alias == 'com.example.App'`?

Comment (andrewazores) [reply src/test/java/io/cryostat/rules/RuleProcessorTest.java]: Oh, I missed updating that spot. Thanks for catching it.

Comment (hareetd) [reply src/test/java/io/cryostat/rules/RuleProcessorTest.java]: No problem

File: src/main/java/io/cryostat/rules/MatchExpressionValidator.java
Comment (hareetd) [inline src/main/java/io/cryostat/rules/MatchExpressionValidator.java:50]: This is for my own understanding of Java, but why doesn't the `validate` method need to state `throws Exception? The reason I'm asking is that in the corresponding unit test, either an `IllegalMatchExpressionException` or a `NashornException` is possible when calling this method.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:50]: `IllegalMatchExpressionException extends RuntimeException`. Java has a distinction between checked and unchecked exceptions. Checked exceptions must be declared thrown and explicitly caught somewhere, but unchecked exceptions don't. NullPointerException is another example of an unchecked exception you've probably seen before - it would really, really suck if NPEs always had to be accounted for with try/catch blocks or throws declarations, because then nearly every line of Java code would need exception handling.

Comment (hareetd) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:50]: Ah makes sense, thanks.

Comment (andrewazores) [reply src/main/java/io/cryostat/rules/MatchExpressionValidator.java:50]: I would say in this case that it probably would make sense for the exception we throw to be a checked kind, since the `TreeVisitor` interface doesn't declare that the methods throw checked exceptions, I can't implement the interface and have mine change the method signatures by adding a new checked exception declaration.

I suppose I could work around this by having the `TreeVisitor` throw an unchecked type, then have the `validator` catch and rethrow those as checked exceptions. That's probably a good idea, actually.

File: src/test/java/io/cryostat/rules/RuleMatcherTest.java
Comment (hareetd) [inline src/test/java/io/cryostat/rules/RuleMatcherTest.java:139]: I think this is unnecessary because it's just a duplicate of the class field `bindings`.

**Review by ebaron — DISMISSED** [review]

**Review by hareetd — DISMISSED** [review]
Reviewed the recent commits as well, looks good to go.

**Review by jan-law — DISMISSED** [review]

**Comment by andrewazores** [comment]
Cleanly rebased, approving reviews were all dismissed so I will merge regardless once status checks pass.

**Review by ebaron — APPROVED** [review]

**Pull request merged**
